scipy/spatial/transform/rotation.py

```diff
@@ -1888,9 +2007,11 @@ def match_vectors(cls, a, b, weights=None):
                              "rotation uniquely.")
 
         kappa = s[0]*s[1] + s[1]*s[2] + s[2]*s[0]
-        sensitivity = ((kappa * np.eye(3) + np.dot(B, B.T)) /
-                       (zeta * a.shape[0]))
-        return cls.from_matrix(C), sensitivity
+        sensitivity = np.mean(weights) / zeta * (
+                kappa * np.eye(3) + np.dot(B, B.T))
+        lambda0 = 0.5 * np.sum(weights * np.sum(b**2 + a**2, axis=1))
+
+        return cls.from_matrix(C), lambda0 - np.sum(s), sensitivity
```

Comment:
I think we should use the weighted root-mean-square distance (RMSD) as the loss function. This is the standard error termThe difference is only a square root.

Rather than calculate `lamdba0` I would just evaluate the weighted RMSD. I don't think there is much efficiency gain from the present approach and is there not a risk that it will return a small negative number if `np.sum(s) > lambda0` due to small numerical errors?

The sensitivity matrix seems like a very specific application. It is also very complicated - I don't understand what it is from the docstring. Should we make it an optional return value, with default False? I would also be fine with removing it altogether.

Reply:
(I'm not sure how to add a review comment to something that wasn't changed, so here is a copy-paste)
```
        zeta = (s[0]+s[1]) * (s[1]+s[2]) * (s[2]+s[0])
        if np.abs(zeta) <= 1e-16:
            raise ValueError("Three component error vector has infinite "
                             "covariance. It is impossible to determine the "
                             "rotation uniquely.")
```
I would make this a warning at most, rather than an exception. If we remove the sensitivity calculation we can remove it entirely.

Reply:
> I think we should use the weighted root-mean-square distance (RMSD) as the loss function. This is the standard error term. The difference is only a square root.

But what's the big difference then? I mean for optimization we definitely use squared differences. I don't want introduce additional concepts to the description (loss function, RMSD = root of loss function).

I'm OK with changing it, but I think there are (very minor) pros and cons.

> Rather than calculate lamdba0 I would just evaluate the weighted RMSD. I don't think there is much efficiency gain from the present approach and is there not a risk that it will return a small negative number if np.sum(s) > lambda0 due to small numerical errors?

If the only concern is the numerical problems  --- I will add 0 as a lower bound for the return. 

Regarding these two points. Initially we didn't think of returning loss/RMSD from the function because it's easy to compute it outside. 

> The sensitivity matrix seems like a very specific application. It is also very complicated - I don't understand what it is from the docstring. Should we make it an optional return value, with default False? I would also be fine with removing it altogether.

It is specific and works only for small perturbation/noise, but I think it's quite important. Basically it gives us some measure of estimation accuracy and it's hard to compute outside of the function (hard to understand too, I admit). It is borderline, but I think the implementation is correct and if someone needs it, he'll be able to make sense out of it. 

I can try to provide an extended example with a Monte-Carlo simulation to clarify this concept (as a part of tutorial).

As for optional return --- a user can just ignore it, I'm not sure it's worth introducing another argument. On the other hand, the same can be said about `return_indices` in `reduce` argument. Should we also remove it? Possibly.

> I would make this a warning at most, rather than an exception. If we remove the sensitivity calculation we can remove it entirely.

Yes, this threshold seems quite arbitrary. I think we should convert it to a warning and consider possible `inf/nan` in sensitivity computations.

Reply:
Sorry for slow communication. I am travelling for a conference. How long do we have left to get this finished?

A very widespread use case for this function is to measure the distance between two structures. The RMSD is the distance used. I think it is unintuitive to return the squared distance rather than the distance itself. For my own use, the first thing I would do after calling the function is to apply a square root to the loss function.

RMSD calculations also form the inner loop of many other algorithms. My concern about the sensitivity matrix is that we are limiting the possible speed we can achieve in the future. I wrote a simple benchmark to test this code:
```
class RotationAlignVectors(Benchmark):

    def setup(self):
        n = 30
        rng = np.random.RandomState(0)
        self.x = rng.normal(size=(n, 3))
        self.y = rng.normal(size=(n, 3))

    def time_align_vectors(self):
        """Time how long it takes to align vectors a million times.
        """
        for i in range(3000):
            Rotation.align_vectors(self.x, self.y)
```
It takes about a second on my laptop. Using a quaternion based code in c++ I can do about 5 million RMSD calculations per second. Making the sensitivity matrix an optional argument means that we can sidestep the time-consuming SVD approach when desired.

Reply:
> Sorry for slow communication. I am travelling for a conference. How long do we have left to get this finished?

No worries, thanks Peter. About 10 days or so I think.

Reply:
>  I think it is unintuitive to return the squared distance rather than the distance itself. For my own use, the first thing I would do after calling the function is to apply a square root to the loss function.

Did you mean *it's not intuitive*? Looks like RMSD = sqrt(2 * loss). Also it's not very clear how to interpret weights in terms of distance. Does it require some explanation?

> Making the sensitivity matrix an optional argument means that we can sidestep the time-consuming SVD approach when desired.

I don't see how Python code can work fast within a loop with a lot of repetitions. From my benchmark with aligning 100 vectors --- SVD and `from_matrix` takes about half time of `align_vectors`. So we are talking about 2x possible speedup, which is not tremendous.

Also as far as I know simple and robust quaternion methods also require SVD (similar to `Rotation.mean`). And I think conceptually only such methods are suitable for `scipy/Python`.

Nothing wrong with optionally disabling, but do we really need it?